File: generator.py
import os
import sys
import shutil
import tempfile
import subprocess
from pathlib import Path

def compile(source:Path, output:Path):
    print(f"Compiling {source} to {output}")
    subprocess.run(["g++", "-Wall", "-std=c++14", "-g", "-o", output.absolute(), source.absolute()])\
        .check_returncode()

class TestGen:

    def __init__(self, filename, generator, solution, outputDir):
        self.filename = filename

        self.outputDir = outputDir
        if os.path.exists(self.outputDir):
            shutil.rmtree(self.outputDir)
        os.mkdir(self.outputDir)

        self.tempDir = Path(tempfile.mkdtemp())

        self.generator = Path(self.tempDir, "generator")
        self.solution = Path(self.tempDir, "solution")
        compile(Path(generator), self.generator)
        compile(Path(solution), self.solution)

        self.test_group = -1
        self.test_in_group = 0
        self.group_list = []
        self.test_list = []

    def End(self):
        print("Summary:")
        cnt = -1
        points = 0
        for ginfo in self.group_list:
            cnt += 1
            points += ginfo[0]
            print(f"\tGroup {cnt:02}: {ginfo[0]} {points:3}\t{ginfo[1]}")
        print(f"TOTAL POINTS: {points}")
        assert(points == 100)
        shutil.rmtree(self.tempDir)

    def NewGroup(self, points, comment = ""):
        if comment is None:
            comment = self.group_list[-1][1] # Previous comment
        self.test_group += 1
        self.test_in_group = 0
        self.group_list.append((points, comment))
        print(f"\nGroup {self.test_group}\n")

    def IncreaseTest(self):
        self.test_in_group += 1

    def GetExtension(self, input):
        ioLetter = 'i' if input else 'o'
        letter = chr(self.test_in_group + ord('a'))
        return f".{ioLetter}{self.test_group:02}{letter}"

    def GetInputFile(self):
        return Path(self.outputDir, self.filename + self.GetExtension(True))

    def GetOutputFile(self):
        return Path(self.outputDir, self.filename + self.GetExtension(False))

    def GenerateAnswer(self, input:Path, output:Path):
        print(f"Generating answer {output}")
        with input.open('r') as finp:
            with output.open('w') as fout:
                subprocess.run([self.solution.absolute()], stdin = finp, stdout = fout,
                               stderr = sys.stdout.buffer) \
                    .check_returncode()

    def StoreTest(self):
        self.test_list.append((self.test_group, self.test_in_group))

    def GenerateTest(self, args):
        self.StoreTest()
        args = [str(arg) for arg in args]
        input = self.GetInputFile()
        print(f"Generating test {input} , args: {args}")
        output = self.GetOutputFile()
        with input.open('w') as finp:
            subprocess.run([str(self.generator)] + args, stdout = finp)\
                .check_returncode()
        self.GenerateAnswer(input, output)
        self.IncreaseTest()

    def GenerateRawTest(self, rawFile):
        self.StoreTest()
        input = self.GetInputFile()
        print(f"Raw test {input}")
        output = self.GetOutputFile()
        input.write_text(rawFile)
        self.GenerateAnswer(input, output)
        self.IncreaseTest()

    def CopyRawTest(self, path):
        self.StoreTest()
        path = Path(path)
        input = self.GetInputFile()
        input.write_bytes(path.read_bytes())
        self.GenerateAnswer(input, self.GetOutputFile())
        self.IncreaseTest()

    def GeneratePointFile(self, pointFilePath:Path):
        lines = [] # (sgroup, egroup, points, comments)
        group_count = -1
        for gr in self.group_list:
            group_count += 1
            if lines and lines[-1][2:] == gr:
                lines[-1] = (lines[-1][0], group_count, gr[0], gr[1])
            else:
                lines.append((group_count, group_count, gr[0], gr[1]))

        with pointFilePath.open('w') as f:
            for l in lines:
                print(f"{l[0]}-{l[1]} {l[2]}{' ' if l[3] else ''}{l[3]}", file=f)

    def GenerateTestDescription(self, output:Path):
        with output.open('w') as f:
            cnt = 0
            print(f"{'Nr':8}\t{'Grupa':5} {'Gr Nr':5} {'GPunkti':8}", file = f)
            for test in self.test_list:
                cnt += 1
                grp = self.group_list[test[0]]
                print(f"{cnt:8}\t{test[0]:5} {test[1]:5} {grp[0]:8}\t{grp[1]}", file = f)

File: src/generator.rs
use anyhow::Result;
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::Command;
use tempfile::tempdir;
use zip::write::FileOptions;

pub fn compile(source: &Path, output: &Path) -> Result<()> {
    println!("Compiling {} to {}", source.display(), output.display());
    let status = Command::new("g++")
        .arg("-Wall")
        .arg("-std=c++14")
        .arg("-g")
        .arg("-o")
        .arg(output)
        .arg(source)
        .status()?;
    if !status.success() {
        anyhow::bail!("g++ failed: {}", status);
    }
    Ok(())
}

#[derive(Debug)]
pub struct TestGen {
    pub filename: String,
    pub output_dir: PathBuf,
    _temp_dir: tempfile::TempDir,
    generator: PathBuf,
    solution: PathBuf,
    boi: bool,
    test_group: i32,
    test_in_group: i32,
    pub group_list: Vec<(i32, String)>,
    pub test_list: Vec<(i32, i32)>,
}

impl TestGen {
    pub fn new(
        filename: &str,
        generator_src: &Path,
        solution_src: &Path,
        output_dir: &Path,
        boi: bool,
    ) -> Result<TestGen> {
        if output_dir.exists() {
            fs::remove_dir_all(output_dir)?;
        }
        fs::create_dir_all(output_dir)?;

        let temp_dir = tempdir()?;
        let generator = temp_dir.path().join("generator");
        let solution = temp_dir.path().join("solution");
        compile(generator_src, &generator)?;
        compile(solution_src, &solution)?;

        Ok(TestGen {
            filename: filename.to_string(),
            output_dir: output_dir.to_path_buf(),
            _temp_dir: temp_dir,
            generator,
            solution,
            boi,
            test_group: -1,
            test_in_group: 0,
            group_list: Vec::new(),
            test_list: Vec::new(),
        })
    }

    pub fn end(self) -> Result<()> {
        println!("Summary:");
        let mut cnt = -1i32;
        let mut points = 0i32;
        for ginfo in &self.group_list {
            cnt += 1;
            points += ginfo.0;
            println!("\tGroup {:02}: {} {}\t{}", cnt, ginfo.0, points, ginfo.1);
        }
        println!("TOTAL POINTS: {}", points);
        if !self.boi {
            assert_eq!(points, 100);
        }
        // temp dir cleaned up when dropped
        Ok(())
    }

    pub fn new_group(&mut self, points: i32, comment: Option<&str>, public: bool) {
        let mut comment_str = comment.map(|s| s.to_string()).unwrap_or_else(|| {
            if let Some(prev) = self.group_list.last() {
                prev.1.clone()
            } else {
                String::new()
            }
        });
        self.test_group += 1;
        self.test_in_group = 0;
        if public {
            comment_str.push_str(" --- PUBLISKA GRUPA");
        }
        self.group_list.push((points, comment_str));
        println!("\nGroup {}\n", self.test_group);
    }

    pub fn increase_test(&mut self) {
        self.test_in_group += 1;
    }

    pub fn get_extension(&self, input: bool, test_id: Option<(i32, i32)>) -> String {
        let test_id = test_id.unwrap_or((self.test_group, self.test_in_group));
        let io_letter = if input { 'i' } else { 'o' };
        if self.boi {
            assert_eq!(test_id.1, 0);
            return format!(".{}{:02}", io_letter, test_id.0);
        }
        let letter = ((test_id.1 as u8) + b'a') as char;
        format!(".{}{:02}{}", io_letter, test_id.0, letter)
    }

    pub fn get_input_file(&self, test_id: Option<(i32, i32)>) -> PathBuf {
        self.output_dir.join(format!(
            "{}{}",
            self.filename,
            self.get_extension(true, test_id)
        ))
    }

    pub fn get_output_file(&self, test_id: Option<(i32, i32)>) -> PathBuf {
        self.output_dir.join(format!(
            "{}{}",
            self.filename,
            self.get_extension(false, test_id)
        ))
    }

    pub fn generate_answer(&self, input: &Path, output: &Path) -> Result<()> {
        println!("Generating answer {}", output.display());
        let fin = std::fs::File::open(input)?;
        let fout = std::fs::File::create(output)?;
        let status = Command::new(&self.solution)
            .stdin(fin)
            .stdout(fout)
            .status()?;
        if !status.success() {
            anyhow::bail!("solution failed: {}", status);
        }
        Ok(())
    }

    pub fn store_test(&mut self) {
        self.test_list.push((self.test_group, self.test_in_group));
    }

    pub fn generate_test<I: AsRef<str>>(&mut self, args: &[I]) -> Result<()> {
        self.store_test();
        let args: Vec<String> = args.iter().map(|s| s.as_ref().to_string()).collect();
        let input = self.get_input_file(None);
        println!("Generating test {} , args: {:?}", input.display(), args);
        let fin = std::fs::File::create(&input)?;
        let status = Command::new(&self.generator)
            .args(&args)
            .stdout(std::process::Stdio::from(fin.try_clone()?))
            .status()?;
        if !status.success() {
            anyhow::bail!("generator failed: {}", status);
        }
        self.generate_answer(&input, &self.get_output_file(None))?;
        self.increase_test();
        Ok(())
    }

    pub fn generate_raw_test(&mut self, raw: &str) -> Result<()> {
        self.store_test();
        let input = self.get_input_file(None);
        println!("Raw test {}", input.display());
        std::fs::write(&input, raw)?;
        self.generate_answer(&input, &self.get_output_file(None))?;
        self.increase_test();
        Ok(())
    }

    pub fn copy_raw_test(&mut self, path: &Path) -> Result<()> {
        self.store_test();
        let input = self.get_input_file(None);
        std::fs::write(&input, std::fs::read(path)?)?;
        self.generate_answer(&input, &self.get_output_file(None))?;
        self.increase_test();
        Ok(())
    }

    pub fn generate_point_file(&self, point_file_path: &Path) -> Result<()> {
        let mut lines: Vec<(i32, i32, i32, String)> = Vec::new();
        let mut group_count = -1i32;
        for gr in &self.group_list {
            group_count += 1;
            if let Some(last) = lines.last_mut() {
                if last.2 == gr.0 && last.3 == gr.1 {
                    last.1 = group_count;
                } else {
                    lines.push((group_count, group_count, gr.0, gr.1.clone()));
                }
            } else {
                lines.push((group_count, group_count, gr.0, gr.1.clone()));
            }
        }
        let mut f = std::fs::File::create(point_file_path)?;
        for l in lines {
            writeln!(
                f,
                "{}-{} {}{}",
                l.0,
                l.1,
                l.2,
                if l.3.is_empty() { "" } else { " " }
            )?;
        }
        Ok(())
    }

    pub fn generate_test_description(&self, output: &Path) -> Result<()> {
        let mut f = std::fs::File::create(output)?;
        writeln!(
            f,
            "{:8}\t{:<5} {:<5} {:<8}",
            "Nr", "Grupa", "Gr Nr", "GPunkti"
        )?;
        let mut cnt = 0i32;
        for test in &self.test_list {
            cnt += 1;
            let grp = &self.group_list[test.0 as usize];
            writeln!(
                f,
                "{:8}\t{:5} {:5} {:8}\t{}",
                cnt, test.0, test.1, grp.0, grp.1
            )?;
        }
        Ok(())
    }

    pub fn generate_test_zip(&self, output: &Path, include_output: bool) -> Result<()> {
        let file = std::fs::File::create(output)?;
        let mut zip = zip::ZipWriter::new(file);
        let options = FileOptions::default().compression_method(zip::CompressionMethod::Stored);
        for test in &self.test_list {
            let input_file = self.get_input_file(Some(*test));
            let mut f = std::fs::File::open(&input_file)?;
            zip.start_file(input_file.file_name().unwrap().to_string_lossy(), options)?;
            std::io::copy(&mut f, &mut zip)?;
            if include_output {
                let output_file = self.get_output_file(Some(*test));
                let mut fo = std::fs::File::open(&output_file)?;
                zip.start_file(output_file.file_name().unwrap().to_string_lossy(), options)?;
                std::io::copy(&mut fo, &mut zip)?;
            }
        }
        zip.finish()?;
        println!(
            "Zipfile {} generated{}.",
            output.display(),
            if !include_output {
                " without output files"
            } else {
                ""
            }
        );
        Ok(())
    }
}
